safe_prime_generator.go

```diff
@@ -0,0 +1,301 @@
+	mutex *sync.Mutex,
+	waitGroup *sync.WaitGroup,
+	rand io.Reader,
+	pBitLen int,
+) {
+	qBitLen := pBitLen - 1
+	b := uint(qBitLen % 8)
+	if b == 0 {
+		b = 8
+	}
+
+	bytes := make([]byte, (qBitLen+7)/8)
+	p := new(big.Int)
+	q := new(big.Int)
+
+	bigMod := new(big.Int)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				waitGroup.Done()
+				return
+			default:
+				_, err := io.ReadFull(rand, bytes)
+				if err != nil {
+					errChan <- err
+					return
+				}
+
+				// Clear bits in the first byte to make sure the candidate has
+				// a size <= bits.
+				bytes[0] &= uint8(int(1<<b) - 1)
+				// Don't let the value be too small, i.e, set the most
+				// significant two bits.
+				// Setting the top two bits, rather than just the top bit,
+				// means that when two of these values are multiplied together,
+				// the result isn't ever one bit short.
+				if b >= 2 {
+					bytes[0] |= 3 << (b - 2)
+				} else {
+					// Here b==1, because b cannot be zero.
+					bytes[0] |= 1
+					if len(bytes) > 1 {
+						bytes[1] |= 0x80
+					}
+				}
+				// Make the value odd since an even number this large certainly
+				// isn't prime.
+				bytes[len(bytes)-1] |= 1
+
+				q.SetBytes(bytes)
+
+				// Calculate the value mod the product of smallPrimes. If it's
+				// a multiple of any of these primes we add two until it isn't.
+				// The probability of overflowing is minimal and can be ignored
+				// because we still perform Miller-Rabin tests on the result.
+				bigMod.Mod(q, smallPrimesProduct)
+				mod := bigMod.Uint64()
+
+			NextDelta:
+				for delta := uint64(0); delta < 1<<20; delta += 2 {
+					m := mod + delta
+					for _, prime := range smallPrimes {
+						if m%uint64(prime) == 0 && (qBitLen > 6 || m != uint64(prime)) {
+							continue NextDelta
+						}
+					}
+
+					if delta > 0 {
+						bigMod.SetUint64(delta)
+						q.Add(q, bigMod)
+					}
+
+					// If `q = 1 (mod 3)`, then `p` is a multiple of `3` so it's
+					// obviously no prime and such `q` should be rejected.
+					// This will happen in 50% of cases and we should detect
+					// and eliminate them early.
+					//
+					// Explanation:
+					// If q = 1 (mod 3) then there exists a q' such that:
+					// q = 3q' + 1
+					//
+					// Since p = 2q + 1:
+					// p = 2q + 1 = 2(3q' + 1) + 1 = 6q' + 2 + 1 = 6q' + 3 =
+					//   = 3(2q' + 1)
+					// So `p` is a multiple of `3`.
+					qMod3 := new(big.Int).Mod(q, big.NewInt(3))
+					if qMod3.Cmp(big.NewInt(1)) == 0 {
+						continue NextDelta
+					}
+
+					// p = 2q+1
+					p.Mul(q, big.NewInt(2))
+					p.Add(p, big.NewInt(1))
+					if !isPrimeCandidate(p) {
+						continue NextDelta
+					}
+
+					break
+				}
+
+				// There is a tiny possibility that, by adding delta, we caused
+				// the number to be one bit too long. Thus we check BitLen
+				// here.
+				if q.ProbablyPrime(20) &&
+					isPocklingtonCriterionSatisfied(p) &&
+					q.BitLen() == qBitLen {
+
+					mutex.Lock()
+					if ctx.Err() == nil {
```

Comment:
Hm, interesting - so if we've closed the `primeChan` then a send to `primeChan` will panic. 
Fortunately, since you're checking the value of `ctx.Done` above, as soon as it receives a signal, it will unblock that channel and immediately return, causing execution to halt.

From running the race detector, I don't think we have a race between where the signal is sent to Done, but we're trying to pass in a value on `primeChan` as it hasn't panic'd on a 100 race runs. So we don't need this check, right?

Reply:
It sometimes panics without this check:
```
$ go test -count 100
panic: send on closed channel

goroutine 5 [running]:
github.com/keep-network/paillier.runGenPrimeRoutine.func1(0x1180ea0, 0xc4200ec100, 0x1180560, 0xc42008e2d0, 0xc4202320a8, 0x1, 0x1, 0xc4200a6180, 0x5, 0xc42000a120, ...)
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:272 +0x462
created by github.com/keep-network/paillier.runGenPrimeRoutine
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:179 +0x22c
exit status 2
FAIL    github.com/keep-network/paillier        0.067s
```

I think it's a similar situation as the one described earlier: https://github.com/keep-network/paillier/pull/8#discussion_r206121820